app/models/user.rb

```diff
@@ -56,6 +68,39 @@ def self.authenticate(login, password)
     nil
   end
 
+  def self.find_or_create_by_provider_email(provider, email, name, provider_id)
+    with_email = User.where(email: email)
+    existing = with_email.find_by(provider: providers[provider]) ||
+               with_email.last
+
+    return existing if existing
+
+    display_name = name || email
+
+    create(
+      login: display_name, # TODO: Consider auto-changing `login` to avoid duplication
+      name: display_name,
+      email: email,
+      provider: providers[provider],
+      provider_id: provider_id.to_s
+    )
```

Comment:
I am wondering if there is a way to set the `city` here? It it used in several places and I am not sure if the app will work with accounts without `city`. Have you tried if everything renders properly without city?